### PR TITLE
Normative: Suppress source text for JSON parse data modified before access

### LINKED ISSUE #35: Assert failed for typedValNode is an ObjectLiteral Parse Node
Consider the example below:
```js
reviver = function(p, v) {
  if (p == "a") {
    this.b = { get x() {return null}, set x(_){throw 666} }
  }
  return v;
}
JSON.parse('{"a":0,"b":1}', reviver);
```
According to the spec, when calling `InternalizeJSONProperty({a: 0, b: { get x() {return null}, set x(_){throw 666} }}, 'b', reviver, 1)`
```js
// According step 5.
val = { get x() {return null}, set x(_){throw 666} }
// According step 6.c
// This will fail, NumberLiteral Parse Node is not ObjectLiteral Parse Node.
Assert(1 is ObjectLiteral Parse Node) 
```

### PR DESCRIPTION
**Pull request by gibson042**
Snapshot the initial data structure before exposing it to ECMAScript code.
Fixes #35
Fixes #39

### CODE REVIEW

**Comment by gibson042** [comment]
@waldemarhorwat @michaelficarra @syg I welcome your reviews.

**Review by michaelficarra — COMMENTED** [review]
`ArrayLiteralContentNodes` returns lists which may contain `Elision` nodes. These are in turn passed to `CreateJSONParseRecord`, followed by `ShallowestContainedJSONValue`. So by my reading, the assertion that is step 2 of `CreateJSONParseRecord` will fail.

File: spec.html [hunk 1]
```diff
@@ -90,13 +91,108 @@ <h1>JSON.parse ( _text_ [ , _reviver_ ] )</h1>
...
+                  1. Let _propName_ be PropName of _candidate_.
+                  1. If SameValue(_propName_, _P_) is *true*, set _propertyDefinition_ to _candidate_.
+                1. Assert: _propertyDefinition_ is not ~empty~.
+                1. NOTE: In the case of JSON text specifying multiple name/value pairs with the same name for a single object (such as <code>{"a":"lost","a":"kept"}</code>), the value for the corresponding property of the resulting ECMAScript object is specified by the last pair with that name.
+                1. Assert: _propertyDefinition_ is <emu-grammar>PropertyDefinition : PropertyName `:` AssignmentExpression</emu-grammar>.
+                1. Let _propertyValueNode_ be the |AssignmentExpression| of _propertyDefinition_.
+                1. Let _entryParseRecord_ be CreateJSONParseRecord(_propertyValueNode_, _P_, ! Get(_val_, _P_)).
+                1. Append _entryParseRecord_ to _entries_.
+          1. Else,
+            1. Assert: _typedValNode_ is not an |ArrayLiteral| Parse Node and not an |ObjectLiteral| Parse Node.
```
Comment (michaelficarra) [inline spec.html:183]: Why is this assertion here? The invariant holds only for this step.

**Comment by michaelficarra** [comment]
Please don't introduce an SDO with the same name as a nonterminal in the grammar (`PropertyDefinitionList`).

File: spec.html [hunk 2]
```diff
@@ -90,13 +91,108 @@ <h1>JSON.parse ( _text_ [ , _reviver_ ] )</h1>
...
+              1. Let _I_ be 0.
+              1. Repeat, while _I_ &lt; _len_,
+                1. Let _propName_ be ! ToString(𝔽(_I_)).
+                1. Let _elementParseRecord_ be CreateJSONParseRecord(_contentNodes_[_I_], _propName_, ! Get(_val_, _propName_)).
+                1. Append _elementParseRecord_ to _elements_.
+                1. Set _I_ to _I_ + 1.
+            1. Else,
+              1. Assert: _typedValNode_ is an |ObjectLiteral| Parse Node.
+              1. Let _properties_ be PropertyDefinitionList of _typedValNode_.
+              1. NOTE: Because _val_ was produced from JSON text and has not been modified, all of its properties are enumerable and all of its property keys are Strings.
```
Comment (michaelficarra) [inline spec.html:169]: You probably also want to mention that this is why they will be in source text order.

**Comment by gibson042** [comment]
> `ArrayLiteralContentNodes` returns lists which may contain `Elision` nodes. These are in turn passed to `CreateJSONParseRecord`, followed by `ShallowestContainedJSONValue`. So by my reading, the assertion that is step 2 of `CreateJSONParseRecord` will fail.

Elisions are not valid in JSON text.

File: spec.html (hunk 1 above)
Comment (gibson042) [reply spec.html:183]: To assert agreement between _parseNode_ and _val_: the latter is an object if and only if the former is an |ArrayLiteral| or |ObjectLiteral|.

File: spec.html (hunk 2 above)
Comment (gibson042) [reply spec.html]: Fixed.

**Comment by gibson042** [comment]
> Please don't introduce an SDO with the same name as a nonterminal in the grammar (`PropertyDefinitionList`).

Do you have an alternative suggestion?

File: spec.html (hunk 1 above)
Comment (syg) [reply spec.html:183]: +1 this assertion is helpful.

File: spec.html [hunk 3]
```diff
@@ -90,13 +91,108 @@ <h1>JSON.parse ( _text_ [ , _reviver_ ] )</h1>
...
+    </ins>
+
     <emu-clause id="sec-internalizejsonproperty" type="abstract operation">
       <h1>
         InternalizeJSONProperty (
           _holder_: an Object,
           _name_: a String,
           _reviver_: a function object,
-          <ins>_valNode_: a Parse Node,</ins>
+          <ins>_parseRecord_: either a JSON Parse Record or *undefined*,</ins>
```
Comment (syg) [inline spec.html:195]:
```suggestion
          <ins>_parseRecord_: either a JSON Parse Record or ~empty~,</ins>
```

And mutadis mutandis below.

**Review by syg — APPROVED** [review]

File: spec.html (hunk 3 above)
Comment (gibson042) [reply spec.html]: Sure, done.

**Pull request merged**